netsim/augment/links.py

```diff
@@ -288,155 +214,290 @@ def augment_link_prefix(link: Box,pools: typing.List[str],addr_pools: Box,link_p
-      ifdata.bridge = link.bridge
-    ifdata.name = link.get("name") or (link[IFATTR][intf_cnt].node + " -> " + link[IFATTR][1-intf_cnt].node)
-    dict_2_update = add_node_interface(ndict[node],ifdata,defaults)
-    value.ifindex = dict_2_update.ifindex
-    interfaces.append(dict_2_update)
-
-  if not 'name' in link:
-    link.name = link_nodes[0].name + " - " + link_nodes[1].name
-
-  for i in range(0,2):
-    link[end_names[i]] = { 'node': link_nodes[i]['name'],'ifname': interfaces[i].get('ifname') }
-
-    remote = link_nodes[1-i].name
-    interfaces[i]['neighbors'] = [{
-        'node': remote,
-        'ifname': interfaces[1-i]['ifname']
-      }]
-    for af in ('ipv4','ipv6'):
-      if af in interfaces[1-i]:
-        interfaces[i]['neighbors'][0][af] = interfaces[1-i][af]
-      if af in interfaces[i]:
-        link[end_names[i]][af] = interfaces[i][af]
+    #
+    # Create node interface data from interfaces attributes augmented with link attributes
+    # and node-relevant link module attributes
+    ifdata = interface_data(
+                link=link,
+                link_attr=link_attr_nomod.union(ndict[node].get('module',[])),
+                ifdata=Box(value))
+    set_interface_name(ifdata,link,intf_cnt)
+    ifdata.pop('node',None)                                       # Remove the node name (not needed within the node)
+    node_intf = add_node_interface(ndict[node],ifdata,defaults)   # Attach new interface to its node
+    value.ifindex = node_intf.ifindex                             # Save ifindex and ifname in link interface data
+    value.ifname  = node_intf.ifname                              # ... needed for things like Graph output module that works with links
+    interfaces.append({ 'node': node, 'data': node_intf })        # Save newly-created interface for the next step
+                                                                  # ... must use dict not Box as Box creates a copy of the data structure
 
-    # JvB: copy module specific link attributes like bgp.local_as
-    mods_with_ifattr = Box({ m : True for m in ndict[remote].get('module',[]) if defaults[m].attributes.get('interface',None) })
-    ifaddr_add_module(interfaces[i]['neighbors'][0],interfaces[1-i],mods_with_ifattr)
+  if common.debug_active('links'):     # pragma: no cover (debugging)
+    print(f'... LAN link data: {link}\n')
 
-  return link
+  # Second phase: build neighbor list from list of newly-created interfaces
+  for node_if in interfaces:
+    ifdata = node_if['data']                                      # Get a pointer to interface data
+    ifdata.neighbors = []
+    node_if['__me'] = True                                        # Set 'that's me' flag so we know which interface to skip
+    for remote_if in interfaces:                                  # Iterate over all interfaces created from this link
+      if '__me' in remote_if:                                     # ... and skip the current interface
```

Comment:
if node_if == remote_if:
  continue
why set a flag?

Reply:
This replaced a very long condition we had in place to deal with loopback links, but I like the suggestion. Will have to use "a is b" though, dicts (or boxes) with identical contents are equal

```
>>> a = { 'x': 1, 'y': 2 }
>>> b = a
>>> c = { 'x': 1, 'y': 2 }
>>> a == c
True
>>> a is c
False
>>> a is b
True
```

Reply:
Fixed the code based on your suggestion. Thanks again!